src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -1,144 +1,218 @@
-
-    default:
-      break;
   }
 
-  if (timeStartRead != 0) {
-    // Apparently we're waiting for some reading.
-    if (timePassedSince(timeStartRead) > SI7021_TIMEOUT) {
-      reset();
-    }
+  // Set the resolution we want
+  if(!setResolution(i2caddr,resolution)){
+    return false;
   }
-  return SI7021_state::New_values == state;
+  
+  //SI7013 has ADC and we need to start the first measurement
+  if (chip_id = CHIP_ID_SI7013){
+    if (!requestADC(i2caddr))
+      return false;
+  }
+
+  //we are now initialized
+  return true;
 }
 
-bool P014_data_struct::getReadValue(float& temperature, float& humidity) {
-  bool success = false;
 
-  switch (state) {
-    case SI7021_state::Uninitialized:
-    {
-      addLog(LOG_LEVEL_INFO, F("SI7021 : sensor not initialized !"));
-      init();
-      break;
-    }
-    case SI7021_state::Initialized:
-    {
-      // No call made to start a new reading
-      // Should be handled in the loop()
-      addLog(LOG_LEVEL_INFO, F("SI7021 : No read started !"));
-      break;
-    }
-    case SI7021_state::Wait_for_temperature_samples:
-    case SI7021_state::Wait_for_humidity_samples:
-    {
-      // Still waiting for data
-      // Should be handled in the loop()
-      addLog(LOG_LEVEL_ERROR, F("SI7021 : Read Error !"));
-      break;
-    }
 
-    case SI7021_state::New_values:
-    {
-      temperature = si7021_temperature / 100.0f;
-      humidity    = si7021_humidity / 10.0f;
-      state       = SI7021_state::Values_read;
-      success     = true;
-
-      if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-        String log = F("SI7021 : Temperature: ");
-        log += temperature;
-        addLogMove(LOG_LEVEL_INFO, log);
-        log  = F("SI7021 : Humidity: ");
-        log += humidity;
-        addLogMove(LOG_LEVEL_INFO, log);
-      }
-      break;
-    }
-    case SI7021_state::Values_read:
-    {
-      // This must be done in a separate call to
-      // make sure we only start reading when the plugin wants us to perform a reading.
-      // Change state of sensor for non blocking reading
-      state = SI7021_state::Initialized;
-      break;
-    }
-  }
-  return success;
+//The main state machine
+//Only perform the measurements with big interval to prevent the sensor from warming up.
+//returns true when the state changes and false otherwise
+bool P014_data_struct::update(uint8_t i2caddr, uint8_t resolution, uint8_t filter_power) {
+  //const unsigned long current_time = millis();
+  //int8_t ret=0;
+  
+  switch(state){
+    case P014_state::Uninitialized:
+        //make sure we do not try to initialize too often
+        //someone could enable the plugin but does not have the hardware connected
+        if (!timeOutReached(last_measurement_time + errCount*SI70xx_INIT_DELAY)) {
+          return false;
+        }
+
+        //we have to stop trying after a while
+        if (errCount>P014_MAX_RETRY){
+            state = P014_state::Error;
+            return true;
+        }
+
```

Comment:
Why not set `last_measurement_time` here?

Reply:
Error does not needs to check for it so to save firmware and cpu cycles.
It is not needed to be set at this point.

Reply:
The point was that it was set in every other code branch following that line.
So now it is set 3 times in the code where it can be set only once, thus reducing the total code size.

Reply:
ok. reduce it from two times to one time in there.